opnode/eth/canonical.go

```diff
@@ -0,0 +1,36 @@
+package eth
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+)
+
+// BlockLinkByNumber Retrieves the *currently* canonical block-hash at the given block-height, and the parent before it.
+// The results of this should not be cached, or the cache needs to be reorg-aware.
+type BlockLinkByNumber interface {
+	BlockLinkByNumber(ctx context.Context, num uint64) (self BlockID, parent BlockID, err error)
+}
+
+type BlockLinkByNumberFn func(ctx context.Context, num uint64) (self BlockID, parent BlockID, err error)
+
+func (fn BlockLinkByNumberFn) BlockLinkByNumber(ctx context.Context, num uint64) (self BlockID, parent BlockID, err error) {
+	return fn(ctx, num)
+}
+
```

Comment:
This is some non-intuitive Go hackery, could we
- make `BlockLinkByNumberFn` private?
- add a comment to the effect that we need to introduce `blockLinkByNumberFn` so that it can implement `BlockLinkByNumber` and let us use an anonymous function a an implementation of `BlockLinkByNumber`?

Reply:
It's a common pattern to define a function type that implements single-function interfaces. Interfaces are more flexible to modify and extend, whereas functions make implementing it really easy.

With most of the RPC interfaces and functions I hope we get different implementations later on; more advanced caching, load-balancing and additional test implementations. E.g. the `l2` package might use it during testing.

Reply:
Keeping the `BlockLinkByNumberFn` public, updated doc to describe the purpose better

Reply:
Necroing b/c I haven't seen this pattern. Interfaces are really nice when you truly don't care about the implementation (`io.Reader` or `io.Writer`), and also when you want a test vs real implementation.

I get splitting up the interface (particularly further down), but I'm curious to see how the type/anonymous function thing works out.

Reply:
https://pkg.go.dev/net/http#HandlerFunc is an example of this pattern in the standard library, and you've probably used it somewhere without realizing already, it's everywhere :)

Reply:
I saw it was being used in another PR and thought maybe it was a common pattern. Does it have a name?
For sure explaining it everywhere is very impractical. If it has a well-known googlable name, e.g. "interaface adapter function type" we could tag uses of the pattern with something like `// interface adapter function type`.

Or we could just let people figure it out I suppose. We do assume a fair amount of Go understanding.

Reply:
"anonymous function interface implementation" I guess? You can find it on google that way, not that well documented though.

Reply:
I like `function to interface adapter` or something along those lines. Important to me is to keep `adapter` and I don't really care for `anonymous` as the function does not need to be anonymous.

I'm fine leaving this as is, but from looking at and playing with it, my gut feeling that it's a little overkill remains. I think the function adapter pattern is really useful when you have a pure function and it avoids having to create an empty struct and define the pure function over that struct, but once you start to rely on closures, I'd just create a struct to implement it.